src/audio/module_adapter/module_adapter.c

```diff
@@ -44,6 +44,55 @@ struct comp_dev *module_adapter_new(const struct comp_driver *drv,
 	return module_adapter_new_ext(drv, config, spec, NULL);
 }
 
+static struct processing_module *module_adapter_mem_alloc(const struct comp_driver *drv,
+							  const struct comp_ipc_config *config)
+{
+	struct comp_dev *dev = comp_alloc(drv, sizeof(*dev));
+
+	if (!dev) {
+		comp_cl_err(drv, "failed to allocate memory for comp_dev");
+		return NULL;
+	}
+
+	/* allocate module information.
+	 * for DP shared modules this struct must be accessible from all cores
+	 * Unfortunately at this point there's no information of components the module
+	 * will be bound to. So we need to allocate shared memory for each DP module
+	 * To be removed when pipeline 2.0 is ready
+	 */
+	int flags = config->proc_domain == COMP_PROCESSING_DOMAIN_DP ?
+			     SOF_MEM_FLAG_USER | SOF_MEM_FLAG_COHERENT : SOF_MEM_FLAG_USER;
+
+	struct processing_module *mod = module_driver_heap_rzalloc(drv->user_heap, flags,
+								   sizeof(*mod));
+
+	if (!mod) {
+		comp_err(dev, "failed to allocate memory for module");
+		goto err;
```

Comment:
Would you consider moving the code from the 'err' label to this location?

Reply:
I'm usually all for "early returns," i.e. for
```
if (fail)
    return;
```
instead of
```
    if (fail)
        goto err;
...
    return 0;
err:
    return -EINVAL;
```
but if there's some cleanup to be done in (all) error cases, I usually prefer a `goto`. When there's only one error case - well, that's a grey zone for me... But since we're likely to extend this function in the future, I'd rather keep it this way.